lib/chef/platform/provider_mapping.rb

```diff
@@ -197,7 +197,8 @@ def platform_provider(platform, version, resource_type)
 
         def resource_matching_provider(platform, version, resource_type)
           if resource_type.kind_of?(Chef::Resource)
-            class_name = resource_type.class.to_s.split('::').last
+            class_name = resource_type.class.name ? resource_type.class.to_s.split('::').last : 
```

Comment:
This should really be `resource_type.class.name`: it won't be different for the classes we are trying to backcompat here, and is more consistent.